File: block_size_test/16x8/module_hd.f90
MODULE module_horizontal_diffusion

!USE module_nc

contains

    subroutine horizontal_diffusion( &
        ids,ide, jds,jde, kds,kde, &
        ims,ime, jms,jme, kms,kme, &
        its,ite, jts,jte, kts,kte )
        IMPLICIT NONE

        !Parameters
        INTEGER ::  ids,ide, jds,jde, kds,kde , &
            ims,ime, jms,jme, kms,kme , &
            its,ite, jts,jte, kts,kte

        !Local variables
        logical :: cf_specified
        logical :: cf_nested
        logical :: cf_open_xs
        logical :: cf_open_xe
        logical :: cf_open_ys
        logical :: cf_open_ye
        logical :: cf_periodic_x
        logical :: cf_polar

        CHARACTER(LEN=1) :: name

        REAL , DIMENSION( ims:ime , kms:kme , jms:jme ) :: field, xkmhd
        REAL , DIMENSION( ims:ime , kms:kme , jms:jme ) :: tendency
        REAL , DIMENSION( ims:ime , kms:kme , jms:jme ) :: tendency_target
        REAL , DIMENSION( ims:ime , kms:kme , jms:jme ) :: tendency_input

        REAL , DIMENSION( ims:ime , jms:jme ) :: mu

        REAL , DIMENSION( ims:ime , jms:jme ) ::    msfux,      &
            msfuy,      &
            msfvx,      &
            msfvx_inv,  &
            msfvy,      &
            msftx,      &
            msfty

        REAL ::     rdx,       &
            rdy,       &
            khdif

        CHARACTER(LEN=50) :: in_fname, out_fname

        integer i,j,k
        integer counter
        integer countDiff
        integer countCalc

        real t_start, t_end
        real temp
        real maxError

        ! Read data from input file
        call cpu_time(t_start)
        in_fname = "inputdata-00001X"
        print *, "Reading input variables from file: ", in_fname
        open(unit = 100, file = in_fname)
        read(100,*) ids, ide, jds, jde, kds, kde
        read(100,*) ims, ime, jms, jme, kms, kme
        read(100,*) its, ite, jts, jte, kts, kte
        read(100,*) cf_specified, cf_nested, &
            cf_open_xs, cf_open_xe, &
            cf_open_ys, cf_open_ye, &
            cf_periodic_x, cf_polar
        read(100,*) name
        read(100,*) field
        read(100,*) tendency
        read(100,*) mu
        read(100,*) msfux, msfuy, msfvx, msfvx_inv, msfvy, msftx, msfty
        read(100,*) khdif, xkmhd, rdx, rdy
        close(100)
        call cpu_time(t_end)
        print '("Time for reading input variables = ",f6.3," seconds.")',t_end-t_start

        !Read variables
        print *, 'Read variables:'
        print *, 'Index d: ', ids, ide, jds, jde, kds, kde
        print *, 'Index m: ', ims, ime, jms, jme, kms, kme
        print *, 'Index t: ', its, ite, jts, jte, kts, kte
        print *, 'Boolean: ', cf_specified, cf_nested, &
            cf_open_xs, cf_open_xe, &
            cf_open_ys, cf_open_ye, &
            cf_periodic_x, cf_polar
        print *, 'Name: ', name
        print *, 'Real: ', khdif, rdx, rdy


        !Call init gpu function
        print *, 'Calling horizontal_diffusion_gpu_init'
        call horizontal_diffusion_gpu_init(0,1,0)

        !Tendency input
!        print *,'Tendency input:'
!        print *, tendency(100:120,1,200)

        !Copy tendency for assesment purpouses
        tendency_input = tendency

        !Call host gpu funciton
        print *, 'Calling horizontal_diffusion_host'
        call horizontal_diffusion_host(ids, ide, jds, jde, kds, kde, &
            ims, ime, jms, jme, kms, kme, &
            its, ite, jts, jte, kts, kte, &
            cf_specified, cf_nested, &
            cf_open_xs, cf_open_xe, cf_open_ys, cf_open_ye, &
            cf_periodic_x, cf_polar, &
            name, &
            field, &
            tendency, &
            mu, &
            msfux, &
            msfuy, &
            msfvx, &
            msfvx_inv, &
            msfvy, msftx, msfty, &
            khdif, &
            xkmhd, &
            rdx, rdy)
        print *, 'Back from horizontal_diffusion_host'

        !Assesing output
        print *, 'Assesing output...'
        call cpu_time(t_start)
        out_fname = "outputdata-00001"
        print *, "Reading output variable from file: ", out_fname
        open(unit = 101, file = out_fname)
        read(101,*) tendency_target
        close(101)
        call cpu_time(t_end)
        print '("Time for reading output variable = ",f6.3," seconds.")',t_end-t_start
        !        print *,'Tendency output-target:'
        !        print *, tendency_target(100:120,1,200)
        !        print *, tendency_target(20:25,35,10)
        !        print *,'Tendency output-gpu:'
        !        print *, tendency(100:120,1,200)
        !        print *, tendency(20:25,35,10)

        !Asses calculation differences between CPU and GPU
!        open(unit = 900, file = 'debug.txt')
!        write(900,*) 'I, K, J, Input, Target, GPU, Diff'
        countDiff = 0
        maxError = 0
!
        DO k = kds, kde
            DO j = jds, jde
                DO i = ids, ide
                    IF( ABS( tendency_target(i,k,j) - tendency(i,k,j) ) > 1.0 ) THEN
                        countDiff = countDiff + 1
!                        write(900,*) i,k,j, tendency_input(i,k,j), tendency_target(i,k,j), tendency(i,k,j), ABS(tendency_target(i,k,j)-tendency(i,k,j))
                    ENDIF
                    !Evaluate max error
                    if (ABS(tendency_target(i,k,j)-tendency(i,k,j)) > maxError) THEN
                        maxError = ABS(tendency_target(i,k,j)-tendency(i,k,j))
                    ENDIF
                ENDDO
            ENDDO
        ENDDO
!        close(900)

        print *,'Big differences between target-tendency and gpu-tendency: ', countDiff, '/', ((ime-ims+1)*(kme-kms+1)*(jme-jms+1))
        !print *,'Calculated values: ', countCalc
        !print *,'Non-calculated values: ', (((ime-ims+1)*(kme-kms+1)*(jme-jms+1)) - countCalc)
        print *,'Max point-to-point error: ', maxError

        !Export output snapshot for comparissons
        print *,'Creating snapshots...'
        call create_snapshot('outputdataGPU.txt', tendency, ims,ime,kms,kme,jms,jme)
        call create_snapshot('outputdataNormal.txt', tendency_target, ims,ime,kms,kme,jms,jme)
        call create_snapshot('outputdataInput.txt', tendency_input, ims,ime,kms,kme,jms,jme)

    END SUBROUTINE horizontal_diffusion

    subroutine create_snapshot(file_name, var, ims,ime,kms,kme,jms,jme)
        CHARACTER(LEN=*) :: file_name
        INTEGER :: ims, ime, jms, jme, kms, kme
        REAL , DIMENSION( ims:ime , kms:kme , jms:jme ) :: var

        !I/O
        open(unit=200, file=file_name, ACTION="write", STATUS="replace")
        !do k=kms,kme
        do j=jms,jme
            write(200, '(435F10.2)')( var(i,1,j) ,i=ims,ime)
        end do
        !end do
        close(200)

    end subroutine create_snapshot


END MODULE module_horizontal_diffusion

program horizontal_diffusion_driver
    use module_horizontal_diffusion
    IMPLICIT NONE

    INTEGER ::   ids, ide, jds, jde, kds, kde, &
        ims, ime, jms, jme, kms, kme, &
        its, ite, jts, jte, kts, kte

    CHARACTER(LEN=80) fname

    print *, "Running program horizontal_diffusion_driver..."

    ! Read data from input file
    fname = "inputdata-00001X"
    print*,'Opening ',trim(fname)
    open(unit = 100, file = fname)
    read(100,*) ids, ide, jds, jde, kds, kde
    read(100,*) ims, ime, jms, jme, kms, kme
    read(100,*) its, ite, jts, jte, kts, kte
    close(100)

    CALL horizontal_diffusion(ids,ide,jds,jde,kds,kde, &
        ims,ime,jms,jme,kms,kme, &
        its,ite,jts,jte,kts,kte)

    stop
end program horizontal_diffusion_driver

File: c_program12.1_multigpu_1gpu_nstreams/module_hd.c
#include <stdio.h>
#include <string.h>
#include "spt1.h"
#include "util.h"

////////////////////////////////////////////////////////////////////////////////////////
// This programs aims to divide the horizontal_diffusion method
// using CPU computing
// as a previous step
// to convert the method to CUDA for using several GPUs

/////////////////////////////////////////////
// Main program
int main() {
	// File for reading input file
	FILE *tp;
	char *fileName = "inputdata-00001X";

	// Model (part 1)
	int i, j, k;
	int ids, ide, jds, jde, kds, kde;
	int ims, ime, jms, jme, kms, kme;
	int its, ite, jts, jte, kts, kte;
	char cf_specified_, cf_nested_, cf_open_xs_, cf_open_xe_, cf_open_ys_, cf_open_ye_,
			cf_periodic_x_, cf_polar_;
	int cf_specified = 0, cf_nested = 0, cf_open_xs = 0, cf_open_xe = 0, cf_open_ys = 0,
			cf_open_ye = 0, cf_periodic_x = 0, cf_polar = 0;
	char name;

	/////////////////////////////////////////////
	// Reading
	printf("Reading parameters from input file %s...\n", fileName);

	startTimer(0);
	tp = fopen(fileName, "rt");

	fscanf(tp, "%d %d %d %d %d %d ", &ids, &ide, &jds, &jde, &kds, &kde);
	fscanf(tp, "%d %d %d %d %d %d ", &ims, &ime, &jms, &jme, &kms, &kme);
	fscanf(tp, "%d %d %d %d %d %d ", &its, &ite, &jts, &jte, &kts, &kte);
	fscanf(tp, "%c %c %c %c %c %c %c %c ", &cf_specified_, &cf_nested_, &cf_open_xs_, &cf_open_xe_,
			&cf_open_ys_, &cf_open_ye_, &cf_periodic_x_, &cf_polar_);
	fscanf(tp, "%c ", &name);

	if (cf_specified_ == 'T')
		cf_specified = 1;
	if (cf_nested_ == 'T')
		cf_nested = 1;
	if (cf_open_xs_ == 'T')
		cf_open_xs = 1;
	if (cf_open_xe_ == 'T')
		cf_open_xe = 1;
	if (cf_open_ys == 'T')
		cf_open_ys = 1;
	if (cf_open_ye_ == 'T')
		cf_open_ye = 1;
	if (cf_periodic_x_ == 'T')
		cf_periodic_x = 1;
	if (cf_polar_ == 'T')
		cf_polar = 1;

	// Model (part 2)
	const int ix = ime - ims + 1;
	const int jx = jme - jms + 1;
	const int kx = kme - kms + 1;
	const int dim3d = ix * jx * kx;
	const int dim2d = ix * jx;
	printf("Number of elements of 3d arrays: %d \n", dim3d);
	printf("Number of elements of 2d arrays: %d \n", dim2d);

	float field[dim3d];
	float tendency[dim3d];
	float xkmhd[dim3d];
	float tendency_target[dim3d];
	float tendency_input[dim3d];

	float mu[dim2d];
	float msfux[dim2d];
	float msfuy[dim2d];
	float msfvx[dim2d];
	float msfvx_inv[dim2d];
	float msfvy[dim2d];
	float msftx[dim2d];
	float msfty[dim2d];

	float rdx, rdy, khdif;

	readArray(tp, field, dim3d);
	readArray(tp, tendency, dim3d);
	readArray(tp, mu, dim2d);
	readArray(tp, msfux, dim2d);
	readArray(tp, msfuy, dim2d);
	readArray(tp, msfvx, dim2d);
	readArray(tp, msfvx_inv, dim2d);
	readArray(tp, msfvy, dim2d);
	readArray(tp, msftx, dim2d);
	readArray(tp, msfty, dim2d);

	readArray(tp, &khdif, 1);
	readArray(tp, xkmhd, dim3d);
	readArray(tp, &rdx, 1);
	readArray(tp, &rdy, 1);

	fclose(tp); // Close file

	// copy tendency -> tendency_input (for comparison purposes)
	memcpy(tendency_input, tendency, sizeof(tendency));

	// Print some read variables
	printf("Indices ids, ide, jds, jde, kds, kde: %4d %4d %4d %4d %4d %4d \n", ids, ide, jds, jde,
			kds, kde);
	printf("Indices ims, ime, jms, jme, kms, kme: %4d %4d %4d %4d %4d %4d \n", ims, ime, jms, jme,
			kms, kme);
	printf("Indices its, ite, jts, jte, kts, kte: %4d %4d %4d %4d %4d %4d \n", its, ite, jts, jte,
			kts, kte);
	printf("Dimensions: %d %d %d \n", ix, jx, kx);
	printf("Flags: %c %c %c %c %c %c %c %c \n", cf_specified_, cf_nested_, cf_open_xs_, cf_open_xe_,
			cf_open_ys_, cf_open_ye_, cf_periodic_x_, cf_polar_);
	printf("Name: %c \n", name);
	printf("Float: rdx rdy: %f %f \n", rdx, rdy);

	stopTimer(0);
	printf("Time for reading input variables: %d ms.\n", getElapsedTimeMs(0));

	/////////////////////////////////////////////
	// Calling horizontal_diffusion host
	printf("Calling horizontal_diffusion method \n");
	startTimer(0);

	horizontal_diffusion_host(&ids, &ide, &jds, &jde, &kds, &kde, &ims, &ime, &jms, &jme, &kms,
			&kme, &its, &ite, &jts, &jte, &kts, &kte, &cf_specified, &cf_nested, &cf_open_xs,
			&cf_open_xe, &cf_open_ys, &cf_open_ye, &cf_periodic_x, &cf_polar, &name, &field,
			&tendency, &mu, &msfux, &msfuy, &msfvx, &msfvx_inv, &msfvy, &msftx, &msfty, &khdif,
			&xkmhd, &rdx, &rdy);

	stopTimer(0);
	printf("Time for running horizontal_diffusion host: %d ms.\n", getElapsedTimeMs(0));

	/////////////////////////////////////////////
	// Assessing output
	printf("Assessing output...\n");
	char *out_fname = "outputdata-00001";
	printf("Reading output variable from file: %s...\n", out_fname);
	tp = fopen(out_fname, "rt");
	readArray(tp, tendency_target, dim3d);
	fclose(tp);

	int countDiff = 0;
	float maxError = 0;

	for (i = 0; i < ix; i++) {
		for (j = 0; j < jx; j++) {
			for (k = 0; k < kx; k++) {
				// Count BIG differences
				if (abs(tendency_target[P3(i, k, j)] - tendency[P3(i, k, j)]) > 1.0) {
					countDiff = countDiff + 1;
				}

				// Calculate max error from target
				if (abs(tendency_target[P3(i, k, j)] - tendency[P3(i, k, j)]) > maxError) {
					maxError = abs(tendency_target[P3(i, k, j)] - tendency[P3(i, k, j)]);
				}

			}
		}
	}

	printf("Big differences between target-tendency and CALC-tendency: %d / %d \n", countDiff,
			dim3d);
	printf("Max point-to-point error: %f \n", maxError);

	//Export output snapshot for comparison
	printf("Creating snapshots...\n");
	create_snapshot("outputdataCalc.txt", tendency, ix, jx, kx);
	create_snapshot("outputdataTarget.txt", tendency_target, ix, jx, kx);
	create_snapshot("outputdataInput.txt", tendency_input, ix, jx, kx);

	printf("Done!\n");
	return 0;
}

